Convert some classes to Kotlin / add tests

csdodd opened the pull request:
Converted some of the concrete classes under the `engine` package to Kotlin.

This included rewriting enums as `sealed` classes, adding null safety, refactoring some code, and finding a couple of bugs.

I have added unit tests for the code that I have changed - this was done before the conversion to Kotlin to ensure that my changes didn't break things.

csdodd commented:
CC @Naitbit 

I appreciate this is a pretty big PR (and completely out of the blue), but I thought you might appreciate it anyway.

I have no problems getting more involved in this repo and bringing it in to a more modern state if there is an interest. If we do that then it'll be easier to maintain / add new features.

Naitbit commented:
Thanks for the PR, and it is good to hear that you would like to get more involved :+1: . We will try to allocate someone to review this PR tomorrow.

michalAngelo commented on app/src/test/java/com/chess/clock/engine/TimeControlTest.kt:
```diff
@@ -0,0 +1,117 @@
+    @Test
+    fun writeToParcelDoesNotThrowException() {
+        val parcel = Mockito.mock(Parcel::class.java)
+        testClass.writeToParcel(parcel, 0)
```
What result it return if method thrown exception?

csdodd replied:
I rewrote these tests to add some proper value. The tests now check that the correct information is being written to `Parcel`

csdodd commented:
I have addressed the comments, and merged upstream conflicts.

A lot of redundant comments and logs have been removed. Variables have been renamed to state their value more clearly.

The Parcel write tests were changed to add more value.
